### PR TITLE
Added remove_row() and remove_col() 

### PR DESCRIPTION
**Pull request by djkato**
Adding functions `remove_col(index: usize)` and `remove_row(index: usize)`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1613,4 +1660,21 @@ r#"[
         let sum_by_col: Vec<u8> = grid.iter_cols().map(|col| col.sum()).collect();
         assert_eq!(sum_by_col, vec![1+4, 2+5, 3+6])
     }
+        #[test]
+    fn remove_col_empty() {
```
Comment (becheran) [inline src/lib.rs:1664]: Remove empty from fn name

File: src/lib.rs [hunk 2]
```diff
@@ -1613,4 +1660,21 @@ r#"[
...
+        let mut grid = grid![[1,2,3,4][5,6,7,8][9,10,11,12][13,14,15,16]];
+        assert_eq![grid.remove_col(3), Some(vec![4, 8, 12, 16])];
+        assert_eq![grid.remove_col(0), Some(vec![1, 5, 9, 13])];
+        assert_eq![grid.remove_col(1), Some(vec![3, 7, 11, 15])];
+        assert_eq![grid.remove_col(0), Some(vec![2, 6, 10, 14])];
+        assert_eq![grid.remove_col(0), None];
+    }
+    #[test]
+    fn remove_row_empty() {
```
Comment (becheran) [inline src/lib.rs:1673]: Same

File: src/lib.rs [hunk 3]
```diff
@@ -1613,4 +1660,21 @@ r#"[
         let sum_by_col: Vec<u8> = grid.iter_cols().map(|col| col.sum()).collect();
         assert_eq!(sum_by_col, vec![1+4, 2+5, 3+6])
     }
+        #[test]
```
Comment (becheran) [inline src/lib.rs:1663]: Thx for adding tests. Could you please also add tests for out of bounds check? For example empty grid or idx greater size?

File: src/lib.rs [hunk 4]
```diff
@@ -559,6 +559,29 @@ impl<T> Grid<T> {
...
+    /// ```
+    /// use grid::*;
+    /// let mut grid = grid![[1,2][3,4][5,6]];
+    /// assert_eq![grid.remove_row(1), Some(vec![3,4])];   
+    /// assert_eq![grid.remove_row(0), Some(vec![1,2])];
+    /// assert_eq![grid.remove_row(0), Some(vec![5,6])];
+    /// assert_eq![grid.remove_row(0), None];
+    /// ```
+    pub fn remove_row(&mut self, row_index: usize) -> Option<Vec<T>> {
+        if self.cols == 0 || self.rows == 0 || row_index > self.rows {
```
Comment (becheran) [inline src/lib.rs:574]: Shouldn't it be >= ?

File: src/lib.rs [hunk 5]
```diff
@@ -589,6 +612,30 @@ impl<T> Grid<T> {
...
+    /// use grid::*;
+    /// let mut grid = grid![[1,2,3,4][5,6,7,8][9,10,11,12][13,14,15,16]];
+    /// assert_eq![grid.remove_col(3), Some(vec![4,8,12,16])];
+    /// assert_eq![grid.remove_col(0), Some(vec![1,5,9,13])];
+    /// assert_eq![grid.remove_col(1), Some(vec![3,7,11,15])];
+    /// assert_eq![grid.remove_col(0), Some(vec![2,6,10,14])];
+    /// assert_eq![grid.remove_col(0), None];
+    /// ```
+    pub fn remove_col(&mut self, col_index: usize) -> Option<Vec<T>> {
+        if self.cols == 0 || self.rows == 0 || col_index > self.cols {
```
Comment (becheran) [inline src/lib.rs:628]: >= ?

**Review by becheran — CHANGES_REQUESTED** [review]
Thx for the pr. Some remarks

**Comment by djkato** [comment]
Thanks for the >=, didn't catch that. Renamed existing tests and added out of bound tests. 

I'd recommend running a formatter on the project. the standard rustfmt makes a huge amount of changes and one can tell it's not  been formatted in a while :).

File: src/lib.rs [hunk 6]
```diff
@@ -589,6 +612,30 @@ impl<T> Grid<T> {
...
+    /// assert_eq![grid.remove_col(0), Some(vec![2,6,10,14])];
+    /// assert_eq![grid.remove_col(0), None];
+    /// ```
+    pub fn remove_col(&mut self, col_index: usize) -> Option<Vec<T>> {
+        if self.cols == 0 || self.rows == 0 || col_index >= self.cols {
+            return None;
+        }
+        let mut residue: Vec<T> = vec![];
+        for i in 0..self.rows {
+			residue.push(self.data.remove(i * self.cols + col_index - i));
```
Comment (becheran) [inline src/lib.rs:633]: Another thing that just came into my mind now is that this is quite inefficient. Every iteration the remaining vector would be reordered for every remove call.

**Review by becheran — CHANGES_REQUESTED** [review]
The removal of cols can be more efficient. Take a look at the pop_col fn. I did some work there to remove the last col more efficient.

File: src/lib.rs (hunk 6 above)
Comment (djkato) [reply src/lib.rs:633]: I played around with optimizing this a bit but sorry I'm relatively new to rust and this is my first github pull request :).
I don't see how every iteration reorders anything, as `push()` appends to the end of the residue vector and `remove()` keeps the order of the original vector. I can't really talk about speed of `remove()` though.
Googling around I found that a faster method to remove the elements would be either `swap_remove()` which reorders the elements, or `retain()`, which would be faster but then I loose the option to return the `Some(residue)`, as:
```rs
let mut i  = 0;
let total_row = self.rows.clone();
self.data.retain(|element: &T| {
    i += 1;
    if ((i- 1)% total_row) == col_index  {
        // residue.push(element.to_owned()); cannot clone reference...
        println!("Removing at {:?}", i - 1);
        return false
    } else {
        return true
    }
});
```
Using a custom loop that splits the vector always at col_index seems more memory intensive as there are multiple new vectors allocated. While the speed of `remove()` isn't the best I don't see another way myself... If you have an idea I'd love to hear it!

Comment (becheran) [reply src/lib.rs:633]: No worries.

The remove function is slow because one has to sort the elements after the removal one position to the left. Doing this in a loop we end up with a runtime complexity of _O(n^2)_. Swap remove is very fast because it is essentially just a swap of last and current element and then decrease the length by one _O(1)_. But, this won't help because it will mess up the order. I think the best possible result is only iterating over all elements once _O(n)_.

What I did in the pop column is using rotations and in the end decrease the size. Anyhow if in doubt about performance it is always good to proof improvements via benchmarks.

You can run the existing benchmark with _cargo bench_.

I can also do this if you want. I just don't know when I will find some time.

Comment (djkato) [reply src/lib.rs:629]: Yeah I understand it's not optimal, I might try to solve it later when I have time. Maybe you'll even have time sooner than me :). I'm already using my version of it though cause I needed it locally. I Made this pull request cause any remove function is better than no function.

Comment (becheran) [reply src/lib.rs]: I pushed a version with uses rotations instead of vec `remove()`. What do you think? The benchmark shows an increase from 120 ms to now only 90 us for a 1000*1000 grid.

Comment (djkato) [reply src/lib.rs]: Not sure I fully understand what's going on there but if it improves performance then great 👌 
Thanks for finding time.

**Review by becheran — APPROVED** [review]
